src/netbase.h

```diff
@@ -64,6 +65,62 @@ struct ProxyCredentials
     std::string password;
 };
 
+/**
+ * List of reachable networks. Everything is reachable by default.
+ */
+class ReachableNets {
+public:
+    void Add(Network net) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        if (net != NET_UNROUTABLE && net != NET_INTERNAL) {
+            LOCK(m_mutex);
+            m_reachable.insert(net);
+        }
+    }
+
+    void Remove(Network net) EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        AssertLockNotHeld(m_mutex);
+        if (net != NET_UNROUTABLE && net != NET_INTERNAL) {
+            LOCK(m_mutex);
+            m_reachable.erase(net);
+        }
+    }
+
+    void RemoveAll() EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        LOCK(m_mutex);
+        m_reachable.clear();
+    }
+
+    bool Contains(Network net) const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
+        LOCK(m_mutex);
+        return m_reachable.count(net) > 0;
+    }
+
+    bool Contains(const CNetAddr& addr) const EXCLUSIVE_LOCKS_REQUIRED(!m_mutex)
+    {
```

Comment:
Added `[[nodiscard]]` and one `AssertLockNotHeld()` where it is not immediately followed by `LOCK()`. I do not see the point in:

```cpp
AssertLockNotHeld(m); // the point of this is to crash if m is held
LOCK(m); // but this will crash if m is held, so what is the point?
```

If it is really desirable to have `AssertLockNotHeld()` before every `LOCK()`, then `AssertLockNotHeld()` should be put inside `LOCK()`.

This is another story:

```cpp
AssertLockNotHeld(m);
if (x) {
    LOCK(m);
    ...
}
```

because if it gets executed with `x == false` and `AssertLockNotHeld()` is missing then it will not catch the problem (by crashing).

Reply:
The double lock detection inside `LOCK()` is only if `DEBUG_LOCKORDER` is defined.

Reply:
Opened https://github.com/bitcoin/bitcoin/pull/27116 to amend the docs.

Reply:
Since https://github.com/bitcoin/bitcoin/pull/27116 is stalled I have added `AssertLockNotHeld()`, e.g.:

```cpp
        AssertLockNotHeld(m_mutex);
        LOCK(m_mutex);
        m_reachable.insert(net);
```